### PR TITLE
Newmac calibration scumonlymac

### PR DESCRIPTION
**Pull request by filmak**
Features of this PR:

- MAC for multi-hop scum networking
- functions for toggling all 16 GPIOs
- skeleton for a scum-only network (currently only two scum)

### CODE REVIEW

File: scm_v3c/applications/freq_sweep_tx_simple/freq_sweep_tx_simple.c [hunk 1]
```diff
@@ -52,4 +53,6 @@ int main(void) {
 //=========================== private =========================================
 void fill_tx_packet(uint8_t *packet, uint8_t packet_len, repeat_rx_tx_state_t state) {
     sprintf(packet, "%d %d %d", state.cfg_coarse, state.cfg_mid, state.cfg_fine);
+		packet[0] = app_vars.index;
```
Comment (daniellovell) [inline scm_v3c/applications/freq_sweep_tx_simple/freq_sweep_tx_simple.c]: Fil says this is not necessary

File: scm_v3c/applications/freq_sweep_tx_simple/freq_sweep_tx_simple.uvprojx [hunk 2]
```diff
@@ -10,12 +10,13 @@
...
       <ToolsetName>ARM-ADS</ToolsetName>
-      <pCCUsed>5060183::V5.06 update 2 (build 183)::ARMCC</pCCUsed>
+      <pCCUsed>5060960::V5.06 update 7 (build 960)::.\ARMCC</pCCUsed>
+      <uAC6>0</uAC6>
       <TargetOption>
         <TargetCommonOption>
           <Device>ARMCM0</Device>
           <Vendor>ARM</Vendor>
-          <PackID>ARM.CMSIS.4.5.0</PackID>
+          <PackID>ARM.CMSIS.5.7.0</PackID>
```
Comment (daniellovell) [inline scm_v3c/applications/freq_sweep_tx_simple/freq_sweep_tx_simple.uvprojx:19]: Will review if this change causes is issues...

File: scm_v3c/radio.c [hunk 3]
```diff
@@ -25,7 +25,7 @@ signed short cdr_tau_history[11] = {0};
 
 //=========================== definition ======================================
 
-#define DIV_ON
+//#define DIV_ON
```
Comment (daniellovell) [inline scm_v3c/radio.c:28]: This should become a configuration parameter of the radio init, #def probably not the best way

File: scm_v3c/radio.c [hunk 4]
```diff
@@ -126,6 +129,7 @@ void        build_TX_channel_table(
 
 // pkt_len should include CRC bytes (add 2 bytes to desired pkt size)
 void send_packet(uint8_t *packet, uint8_t pkt_len) {
+		gpio_12_set();
```
Comment (daniellovell) [inline scm_v3c/radio.c]: Remove GPIO toggles inside radio -Fil

File: scm_v3c/radio.c [hunk 5]
```diff
@@ -179,20 +196,26 @@ void cb_endFrame_rx_radio(uint32_t timestamp){
...
-    
-    if(packet_len == radio_vars.rxPacket_len && radio_getCrcOk()) {
+    //if(packet_len == radio_vars.rxPacket_len && radio_getCrcOk()) {
+		if(radio_getCrcOk()) {
         // Only record IF estimate, LQI, and CDR tau for valid packets
         radio_vars.IF_estimate        = radio_getIFestimate();
         radio_vars.LQI_chip_errors    = radio_getLQIchipErrors();
-        
         radio_vars.radio_rx_cb(radio_vars.rxPacket, packet_len);
+				gpio_14_toggle();
```
Comment (daniellovell) [inline scm_v3c/radio.c]: Remove

File: scm_v3c/radio.c [hunk 6]
```diff
@@ -151,8 +156,19 @@ void receive_packet(uint8_t pkt_len) {
     radio_rxNow();
     rftimer_setCompareIn(rftimer_readCounter() + TIMER_PERIOD_RX);
     radio_vars.receiveDone = false;
-
+		//printf("rftimer set %d to %d \r\n", rftimer_readCounter(), rftimer_readCounter() + TIMER_PERIOD_RX);
     while (radio_vars.receiveDone==false);
+		//printf("rftimer released\r\n");
+}
+
+void receive_packet_indefinitely(uint8_t pkt_len) {
```
Comment (daniellovell) [inline scm_v3c/radio.c]: Document that its a blocking wait for a packet

File: scm_v3c/rftimer.c [hunk 7]
```diff
@@ -113,7 +113,7 @@ void rftimer_disable_interrupts(void){
 void rftimer_disable_interrupts_by_id(uint8_t id){
     // disable compare interrupt
     *RF_TIMER_REG_CONTROL_ADDRESES[id] = 0x0000;
-    ICER = 0x80;
+    //ICER = 0x80;
```
Comment (daniellovell) [inline scm_v3c/rftimer.c]: Is this ok? Needs testing

File: scm_v3c/scm3c_hw_interface.c [hunk 8]
```diff
@@ -14,8 +14,9 @@
 #define DAC_2M_SETTING_LEN          5
 
 // initialized value for frequency configuration
-#define INIT_HF_CLOCK_FINE          17
```
Comment (daniellovell) [inline scm_v3c/scm3c_hw_interface.c]: Remove changes of board-specific values

File: scm_v3c/radio.c [hunk 9]
```diff
@@ -179,20 +196,26 @@ void cb_endFrame_rx_radio(uint32_t timestamp){
         &radio_vars.rxpk_rssi,
         &radio_vars.rxpk_lqi
     );
-        
-    radio_rfOff();
-    
-    if(packet_len == radio_vars.rxPacket_len && radio_getCrcOk()) {
+    //if(packet_len == radio_vars.rxPacket_len && radio_getCrcOk()) {
```
Comment (daniellovell) [inline scm_v3c/radio.c]: Proposal to have receive_packet_length(uint8_t packet_len) and receive_packet(void) where the latter does not know a priori the length to expect and does not check it.

File: scm_v3c/applications/freq_sweep_rx/freq_sweep_rx.c [hunk 10]
```diff
@@ -180,7 +180,7 @@ void    cb_endFrame_rx(uint32_t timestamp){
     radio_rfOff();
     
     if(
-        app_vars.packet_len == LEN_RX_PKT && (radio_getCrcOk())
```
Comment (daniellovell) [inline scm_v3c/applications/freq_sweep_rx/freq_sweep_rx.c]: Restore packet_len check

**Review by daniellovell — CHANGES_REQUESTED** [review]

File: scm_v3c/applications/freq_sweep_tx_simple/freq_sweep_tx_simple.uvprojx (hunk 2 above)
Comment (daniellovell) [reply scm_v3c/applications/freq_sweep_tx_simple/freq_sweep_tx_simple.uvprojx:19]: This is fine as far as I can tell in my testing. My Keil automatically changes the pack ID to ARM.CMSIS.5.3.0, so I had no issues building.

File: scm_v3c/scm3c_hw_interface.c (hunk 8 above)
Comment (daniellovell) [reply scm_v3c/scm3c_hw_interface.c]: Fixed in e6eeb40567b76b6e8c1851f27788de37049cf0a9

File: scm_v3c/radio.c (hunk 4 above)
Comment (daniellovell) [reply scm_v3c/radio.c]: Done in 8cc73b30b98c561e0eee2eb3df5beb06b6cf01e1

File: scm_v3c/rftimer.c (hunk 7 above)
Comment (daniellovell) [reply scm_v3c/rftimer.c]: Addressed by a0b02cebde38e6f564c9355ee45211c3c370a974

File: scm_v3c/radio.c (hunk 9 above)
Comment (daniellovell) [reply scm_v3c/radio.c]: Implemented in dd1e918347eb80e30ab1b1c1f82230a4d2b77e00

File: scm_v3c/applications/freq_sweep_tx_simple/freq_sweep_tx_simple.c (hunk 1 above)
Comment (filmak) [reply scm_v3c/applications/freq_sweep_tx_simple/freq_sweep_tx_simple.c]: Yes, reverting this change is fine. I was using the index to do PDR testing. What I should really do is make a new project explicitly for PDR testing.

File: scm_v3c/radio.c (hunk 9 above)
Comment (filmak) [reply scm_v3c/radio.c]: In my protocols I effectively rewrite "receive packet" and the "endframe_cb" so any changes are fine. I believe that many of the radio example applications do use these. if I'm understanding the changes correctly, they should work.

File: scm_v3c/rftimer.c (hunk 7 above)
Comment (filmak) [reply scm_v3c/rftimer.c]: These changes do not seem to influence the operation of my protocols. I'm not sure if it's necessary, but I individually enable_by_id each RFTimer compare interrupt.

**Pull request merged**